op-proposer/proposer/driver.go

```diff
@@ -269,18 +279,62 @@ func (l *L2OutputSubmitter) FetchL2OOOutput(ctx context.Context) (*eth.OutputRes
 	return output, true, nil
 }
 
-// FetchDGFOutput gets the next output proposal for the DGF.
+// FetchDGFOutput queries the DGF for the latest game and infers whether it is time to make another proposal
+// If necessary, it gets the next output proposal for the DGF, and returns it along with
+// a boolean for whether the proposal should be submitted at all.
 // The passed context is expected to be a lifecycle context. A network timeout
 // context will be derived from it.
-func (l *L2OutputSubmitter) FetchDGFOutput(ctx context.Context) (*eth.OutputResponse, error) {
-	ctx, cancel := context.WithTimeout(ctx, l.Cfg.NetworkTimeout)
+func (l *L2OutputSubmitter) FetchDGFOutput(ctx context.Context) (*eth.OutputResponse, bool, error) {
+	cCtx, cancel := context.WithTimeout(ctx, l.Cfg.NetworkTimeout)
 	defer cancel()
+	callOpts := &bind.CallOpts{
+		From:    l.Txmgr.From(),
+		Context: cCtx,
+	}
 
-	blockNum, err := l.FetchCurrentBlockNumber(ctx)
+	gameCount, err := l.dgfContract.GameCount(callOpts)
 	if err != nil {
-		return nil, err
+		return nil, false, fmt.Errorf("could not query DisputeGameFactory.gameCount(): %w", err)
 	}
-	return l.FetchOutput(ctx, blockNum)
+
+	if gameCount.Sign() == 1 {
+		// If there is at least one game, ensure
+		// enough time has lapsed that we need to make another proposal.
+
+		latestGameIndex := new(big.Int).Sub(gameCount, big.NewInt(1))
+		latestGame, err := l.dgfContract.GameAtIndex(callOpts, latestGameIndex)
+		if err != nil {
+			return nil, false, fmt.Errorf("could not query DisputeGameFactory.GameAtIndex(%d): %w", latestGameIndex, err)
+		}
+
+		timeSinceLastGame := time.Since(time.Unix(int64(latestGame.Timestamp), 0))
+		if timeSinceLastGame <= l.Cfg.ProposalInterval {
+			l.Log.Info("Duration since last game not past proposal interval", "duration", timeSinceLastGame)
+			return nil, false, nil
+		}
```

Comment:
This is a potential DoS vector. The proposer needs to base its decision on when to propose on the last game _it_ created, not just the last game created.  Otherwise I can publish and invalid output root every 59 minutes, counter it in the same tx so I don't lose my bond and `op-proposer` will never propose a valid output root.

Reply:
The address that created the game is available as `claims[0].Claimant` from the `DisputeGameFactory` created for the game (which is at the address `latestGame.Proxy`).

Reply:
Thanks for flagging, we were aware of something like this but wanted to harden the logic towards a multi-proposer and challenger scenario in a follow up. But since it seems very easy to just use the `Claimant` field, we can probably do it in this PR already @geoknee

Reply:
Thanks for flagging. 

The complexity of this function is going to increase quite a bit if we need to find the latest game `op-proposer` created on chain. Perhaps we could search for the latest tx from `op-proposer`, that would be a bit simpler.

I am wondering about reverting to a design we had on an earlier version of this PR, where `op-proposer` proposes:
* immediately on startup
* every 1hr after that
without doing any on-chain checks.  The main downside of this design is we may get a few more proposals than desired if the proposer is restarted often. The upside is that it is very simple.

We had moved away from that design because we thought the desired behaviour was "ensure there is one proposal per hour _globally_ (i.e. by any proposer)" -- looks like that isn't the case however.

Reply:
I think it's reasonably straight-forward to implement. I'd still split it out into a separate function to make the code clear but it's just a matter of iterating backwards through the list of games until you either find a game that's more than an hour ago (you should propose) or a game that came from the proposer's claimant (you should not propose).

Reply:
We should not go back to proposing immediately after startup. That defeats the purpose of uniform proposal intervals across restarts.

I think what Adrian proposes is an algo like:
iterate backwards over latest games
1. if it's older than 1h, propose
2. if claimant is not this proposer, continue to next game
3. if claimant is this proposer, don't propose (we are now in the <1h case, due to 1.)
4. next iteration will check next older game

Do you think we can still do this as part of this PR @geoknee ?

Reply:
Yes this is doable. I think there may be an edge case where someone proposes during the loop where we iterate over games. It can all be handled, I am sure.

Reply:
Unless it's us proposing we can not worry about anyone else (and we aren't proposing because we're busy running this loop) - we just grab the game count at the start and iterate backwards from there.  The game array is append only so any new games won't affect us.  An L1 reorg might change things but that's very unlikely given the tx mgr waits for a number of confirmations.

Reply:
So my suggestion would be to switch over how we do contract bindings as it just makes this kind of thing easier and comes out with neater code. Took a stab at it with https://github.com/ethereum-optimism/optimism/pull/11472 and I can clean it up and add the tests tomorrow if it's useful.  It does increase the scope a bit (could potentially convert to the new contract style first and then do the timing changes if you want to keep things smaller).

Its returning more stuff than is ideal at the moment as well to accomodate the case where there are no dispute games so if we can find a different solution for that it will simplify things some more too.

Reply:
Thanks Adrian, your approach looks like a good path forward so if you're happy to carry that on then so am I!

Reply:
ok cleaned that up and added tests - I've tagged you and @sebastianst for review.  I suspect we should just merge that PR into this then finish any remaining details here and merge to develop.

Reply:
@ajsutton looks great, left a review.

Reply:
Merged #11472 into this one so we now find the last proposal we made of the game type we're about to propose.